log_utils/get_logger.py

```diff
@@ -3,26 +3,111 @@
-LOG_HANDLER.setFormatter(LOG_FORMATTER)
+class LogFilter(logging.Filter):
+    def __init__(self, priv_vals, mask='[PRIVATE VALUE HIDDEN]'):
+        self.priv_vals = priv_vals
+        self.mask = mask
+        logging.Filter.__init__(self)
 
+    def filter(self, record):
+        for priv_val in self.priv_vals:
+            record.msg = record.msg.replace(priv_val, self.mask)
 
-def get_logger(name):
+        if "InvalidAccessKeyId" in record.msg:
+            record.msg = (
+                'Whoops, our S3 keys were rotated during your '
+                'build and became out of date. This was not a '
+                'problem with your site build, but if you restart '
+                'the failed build it should work on the next try. '
+                'Sorry for the inconvenience!'
+            )
+
+        return len(record.msg) > 0
+
+
+class Formatter(logging.Formatter):
+    '''
+    A more forgiving formatter that will fill in blank values if our custom
+    attributes are missing
+    '''
+    def __init__(self, keys, *args, **kwargs):
+        self.keys = keys
+        logging.Formatter.__init__(self, *args, **kwargs)
+
+    def format(self, record):
+        '''
+        Add missing values before formatting as normal
+        '''
+        for key in self.keys:
+            if (key not in record.__dict__):
+                record.__dict__[key] = ''
+
+        return super().format(record)
+
+
+class StreamToLogger:
+    def __init__(self, logger, log_level=logging.INFO):
+        self.logger = logger
+        self.log_level = log_level
+
+    def write(self, buf):
+        for line in buf.rstrip().splitlines():
+            self.logger.log(self.log_level, line.rstrip())
+
+    def flush(self):
+        pass
+
+
+def get_logger(name, attrs={}):
     '''
     Gets a logger instance configured with our formatter and handler
     for the given name.
     '''
     logger = logging.getLogger(name)
-    logger.setLevel(os.environ.get('LOGLEVEL', logging.INFO))
-    logger.addHandler(LOG_HANDLER)
-    return logger
+
+    return logging.LoggerAdapter(logger, attrs)
+
+
+def init_logging(private_values, attrs={}, db_url=None, skip_logging=False):
+    date_fmt = '%Y-%m-%d %H:%M:%S'
+    style_fmt = '{'
+    short_fmt = '{asctime} {levelname} [{name}] {message}'
+    long_fmt = '{asctime} {levelname} [{name}] '
+    for key in attrs.keys():
+        long_fmt = long_fmt + '@' + key + ': {' + key + '} '
+    long_fmt = long_fmt + '@message: {message}'
+
+    extra_attrs = attrs.keys()
+
+    log_filter = LogFilter(private_values)
+
+    log_level = logging.INFO
+
+    stream_formatter = Formatter(extra_attrs, long_fmt, date_fmt, style_fmt)
+
+    stream_handler = logging.StreamHandler(sys.stdout)
+    stream_handler.setFormatter(stream_formatter)
+    stream_handler.setLevel(log_level)
+    stream_handler.addFilter(log_filter)
+
+    handlers = [stream_handler]
+
+    if not skip_logging and db_url:
+        buildId = attrs['buildid']
```

Comment:
buildid .... should this be "buildId" or "build_id"?

Reply:
i'll make it `build_id` for consistency